src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,24 +1,137 @@
 package core.basesyntax;
 
-/**
- * <p>Реалізувати свою HashMap, а саме методи `put(K key, V value)`, `getValue()` та `getSize()`.
- * Дотриматися основних вимог щодо реалізації мапи (initial capacity, load factor, resize...)
- * За бажанням можна реалізувати інші методи інтрефейсу Map.</p>
- */
+import java.util.Objects;
+
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+
+    private Node<K, V>[] table;
+    private float loadFactor;
+    private int size;
+
+    public MyHashMap() {
+        table = (Node<K, V>[]) new Node[DEFAULT_CAPACITY];
+        loadFactor = DEFAULT_LOAD_FACTOR;
+    }
 
     @Override
     public void put(K key, V value) {
+        if (key == null) {
+            putForNullKey(value);
+            return;
+        }
 
+        int index = indexFor(hash(key.hashCode()), table.length);
+        Node<K, V> node = table[index];
+        while (node != null) {
+            if (key.equals(node.key)) {
+                node.value = value;
+                return;
+            }
+            node = node.next;
+        }
+
+        addNode(key, value, index);
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            Node<K, V> node = table[0];
+            while (node != null) {
+                if (key == node.key) {
+                    return node.value;
+                }
+                node = node.next;
+            }
+        }
+
+        int index = indexFor(hash(key.hashCode()), table.length);
+        Node<K, V> node = table[index];
+        while (node != null) {
+            if (Objects.equals(key, node.key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
+
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int hash(int h) {
```

Comment:
what's happening inside this method, why do we need it, can we skip it?

Reply:
As I understood we need this method for more optimized allocation nodes to the buckets. But I don't understand clearly how is it done (I wrote about it above in answer to Leonid coment). So I used an example from the article, that was given to us as a theory to HashMap topic.
I refactored this method to its variant in the original HashMap.

Reply:
About putForNullKey(). It was described in the article you gave me as theory to HashMap topic. Also as the implementation of method hash(). So please give me theory that will be actual because it confuses. 
The same situation was with deprecated article about LinkedList. I lost a lot of time trying to uderstand why my tests fail.